### PR TITLE
[Resources.Azure] Add support for Container App Jobs

### LINKED ISSUE #2036: Support for Azure Container Apps Jobs
### Component

OpenTelemetry.Resources.Azure

### Is your feature request related to a problem?

In https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1565, an `AzureContainerAppsResourceDetector` for Azure Container Apps (ACA) was added.

It works well for normal ACA resources, but seems not to support [ACA **Jobs**](https://learn.microsoft.com/en-us/azure/container-apps/jobs?tabs=azure-cli).

In my application insights, the normal container app is recognized but the program running in an ACA Job is shown as `unknown_service:dotnet`.

### What is the expected behavior?

ACA Job resources are detected correctly similar to regular ACA apps, i.e. the `service.name`, `service.version`, `service.instance.id`, `cloud.provider` and `cloud.platform` properties are set.

### Which alternative solutions or features have you considered?

A temporary workaround could be - setting these properties manually.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by hansmbakker**
Fixes #2036
Design discussion issue #

## Changes

This adds support for the detection of [Container App Jobs](https://learn.microsoft.com/en-us/azure/container-apps/jobs?tabs=azure-cli) and fixes the issue that normal Container App are recognized but that containers running as a ACA Job are shown as `unknown_service:dotnet`.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by Kielek** [comment]
Please update CHANGELOG.md and README.md.

I will leave logic review for @rajkumar-rangaraj and @TimothyMothra.

**Comment by hansmbakker** [comment]
> Please update CHANGELOG.md and README.md.
@Kielek done

File: src/OpenTelemetry.Resources.Azure/ResourceAttributeConstants.cs
Comment (TimothyMothra) [inline src/OpenTelemetry.Resources.Azure/ResourceAttributeConstants.cs:31]: nit: needs a comment labeling these variables, same as the other groups.
```suggestion
    // Azure Container Apps Jobs environment variables
    internal const string AzureContainerAppJobNameEnvVar = "CONTAINER_APP_JOB_NAME";
    internal const string AzureContainerAppJobExecutionNameEnvVar = "CONTAINER_APP_JOB_EXECUTION_NAME";
```

File: test/OpenTelemetry.Resources.Azure.Tests/AzureResourceDetectorTests.cs
Comment (TimothyMothra) [inline test/OpenTelemetry.Resources.Azure.Tests/AzureResourceDetectorTests.cs:136]: I understand that @hansmbakker didn't introduce the try/catch, this was copied from the structure of the above test.

@rajkumar-rangaraj, any idea why this is wrapped in a try/catch? If either of these Sets fail, not only would the test fail but it would be very difficult to identify why.

**Comment by TimothyMothra** [comment]
Hi @hansmbakker, thanks for the contribution. Were you able to test this in a sample app to confirm that this Environment Variable exists?

@rajkumar-rangaraj, do we need to get this reviewed from anyone on the service team to confirm we can take a dependency on this environment variable? Do we need to share this with the other languages to make the same change?

**Comment by hansmbakker** [comment]
@TimothyMothra I created a sample app and added it in the last commit. I connected it to application insights.

Application map (note that it now correctly shows the job name instead of `unknown_service:dotnet`):
image

Example activity making a call to microsoft.com and throwing an example exception:
image

Environment variables:
image

**Comment by hansmbakker** [comment]
> do we need to get this reviewed from anyone on the service team to confirm we can take a dependency on this environment variable

@TimothyMothra if you're reaching out to the service team anyway, it would be good to request more useful environment variables - it would be nice to also log the image name and image tag for example.

File: examples/Azure/Example.ContainerAppJob/IHostApplicationBuilderExtensions.cs
Comment (Kielek) [inline examples/Azure/Example.ContainerAppJob/IHostApplicationBuilderExtensions.cs:52]: Is there any reason why cannot use .`ConfigureResource(builder => builder.AddAzureContainerAppsResourceDetector())`?

It could remove all magic related to InternalsVisibleTo.

What is more, I think it is part which should not be included into documentation.

**Review by Kielek — CHANGES_REQUESTED** [review]

File: examples/Azure/Example.ContainerAppJob/IHostApplicationBuilderExtensions.cs
Comment (cijothomas) [reply examples/Azure/Example.ContainerAppJob/IHostApplicationBuilderExtensions.cs]: > extension methods only work on a specific builder type.

I am not sure I follow that. ResourceDetectors provide an extension method on `this ResourceBuilder builder` only, and it works with both way of initializing SDK, be it SDK.Create* or builder.Services.OpenTelemetry() style! It is up to app owner to decide which way they should initialize SDK, but the person authoring ResourceDetector should be un-affected by that choice, as they *always* provide extension method on `this ResourceBuilder` only.

> From a developer perspective, I found it confusing and annoying at least 🤦🏼‍♂️.

I am sorry you feel that way! That was not the intend, and based on my above understanding (shared above), that is not the case either. But if there is some bug which prevent this from working, we can fix that. Also, If https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/customizing-the-sdk/README.md#resource doc is not making this clear, I think we can add some wording to avoid any confusions there.

Comment (cijothomas) [reply examples/Azure/Example.ContainerAppJob/IHostApplicationBuilderExtensions.cs]: Given example app is removed, resolving this conversation. Feel free to re-open if there are concerns/questions.

Comment (hansmbakker) [reply examples/Azure/Example.ContainerAppJob/IHostApplicationBuilderExtensions.cs]: I'm sorry, I only saw
- the `OpenTelemetryBuilder` returned from `builder.Services.AddOpenTelemetry()`
- the `TracingProviderBuilder` inside `OpenTelemetryBuilder.WithTracing(...)`
- the `MeterProviderBuilder` inside `OpenTelemetryBuilder.WithMetrics(...)`

I missed the point that `ConfigureResource(...)` (which does indeed give access to the `ResourceBuilder` needed for `AddAzureContainerAppsResourceDetector`) is usable on `builder.Services.AddOpenTelemetry()` as well and not only on Sdk.CreateTracerProviderBuilder() - my apologies.

File: src/OpenTelemetry.Resources.Azure/CHANGELOG.md
Comment (Kielek) [inline src/OpenTelemetry.Resources.Azure/CHANGELOG.md:7]:
```suggestion
* Added support for [Azure Container Apps jobs](https://learn.microsoft.com/en-us/azure/container-apps/jobs?tabs=azure-cli).
  ([#2064](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2064))

```

File: test/OpenTelemetry.Resources.Azure.Tests/AzureResourceDetectorTests.cs
Comment (hansmbakker) [reply test/OpenTelemetry.Resources.Azure.Tests/AzureResourceDetectorTests.cs:136]: @rajkumar-rangaraj could you please pick this up? I indeed followed the structure from the `AzureContainerAppsResourceDetectorReturnsResourceWithAttributes()` test.

(I cannot finish this)

Comment (cijothomas) [reply test/OpenTelemetry.Resources.Azure.Tests/AzureResourceDetectorTests.cs:136]: This is not a blocker for merging the PR, so don't worry about this. (Yes, we can follow up separately to address Mothra's comment.)

Comment (TimothyMothra) [reply test/OpenTelemetry.Resources.Azure.Tests/AzureResourceDetectorTests.cs:136]: Yes, please don't block on this. I was just asking questions.

Comment (hansmbakker) [reply test/OpenTelemetry.Resources.Azure.Tests/AzureResourceDetectorTests.cs:136]: I resolved the comment

**Review by rajkumar-rangaraj — CHANGES_REQUESTED** [review]
I’ve engaged the Azure Container Apps team to confirm whether we can reliably use the environment variables `CONTAINER_APP_JOB_NAME` and `CONTAINER_APP_JOB_EXECUTION_NAME`. Until we receive confirmation, I recommend putting this PR on hold.

**Comment by cijothomas** [comment]
> I’ve engaged the Azure Container Apps team to confirm whether we can reliably use the environment variables `CONTAINER_APP_JOB_NAME` and `CONTAINER_APP_JOB_EXECUTION_NAME`. Until we receive confirmation, I recommend putting this PR on hold.

I am okay to merge the PR, given this package is experimental anyway. If we chose to merge now, lets add an explicit warning to the ACAJobs methods/doc that this is highly experimental and subject to even more change (or similar wording to indicate the unstability part).

@hansmbakker do you prefer to merge now (with a additional doc warning suggested above) OR wait for an official confirmation from ACA team on this?

**Comment by TimothyMothra** [comment]
> I am okay to merge the PR, given this package is experimental anyway.

I'm in favor of waiting before we merge. Users aren't blocked and can write their own resource detector for the short term.

@rajkumar-rangaraj do you have any estimate on how long till we get a response from the server team? As this week is Hackathon, I wouldn't expect a response this week.

@cijothomas would you be comfortable waiting 2-3 weeks for a response? I'd like to avoid extra churn if possible :)

**Comment by hansmbakker** [comment]
> @hansmbakker do you prefer to merge now (with a additional doc warning suggested above) OR wait for an official confirmation from ACA team on this?
@cijothomas  I'm ok with waiting for a confirmation if it's a reasonable timeframe 👍🏼

**Review by rajkumar-rangaraj — APPROVED** [review]
LGTM

I received confirmation that we can rely on these environment variables for ACA jobs.

**Review by Kielek — APPROVED** [review]

**Review by TimothyMothra — APPROVED** [review]

**Comment by Kielek** [comment]
@hansmbakker, tests are failing. Could you please fix it?

**Comment by hansmbakker** [comment]
@Kielek the app name / job name env vars were not cleared in between tests which caused the test failure.

**Review by Kielek — APPROVED** [review]

**Pull request merged**
